In [8]:
!pip install reed

In [9]:
# hide
import re
import json
from html.parser import HTMLParser

import numpy as np

In [10]:
# hide
class MyHTMLParser(HTMLParser):

    def __init__(self):
        self.string = ''
        super().__init__()

    def handle_data(self, data):
        self.string = self.string + ' ' + data
        return (data)

    def return_data(self):
        return self.string.strip().replace('  ', ' ')

In [11]:
# hide
from googleapiclient.discovery import build
import io, os
from googleapiclient.http import MediaIoBaseDownload
from google.colab import auth

auth.authenticate_user()

drive_service = build('drive', 'v3')
results = drive_service.files().list(
        q="name = 'creds.json'", fields="files(id)").execute()
creds = results.get('files', [])
results = drive_service.files().list(
        q="name = 'job_data.csv'", fields="files(id)").execute()
data = results.get('files', [])

filename = "/content/.reed/creds.json"
os.makedirs(os.path.dirname(filename), exist_ok=True)

request = drive_service.files().get_media(fileId=creds[0]['id'])
fh = io.FileIO(filename, 'wb')
downloader = MediaIoBaseDownload(fh, request)
done = False
while done is False:
    status, done = downloader.next_chunk()
    print("Download %d%%." % int(status.progress() * 100))
os.chmod(filename, 600)

filename = "/content/data/job_data.csv"
os.makedirs(os.path.dirname(filename), exist_ok=True)

request = drive_service.files().get_media(fileId=data[0]['id'])
fh = io.FileIO(filename, 'wb')
downloader = MediaIoBaseDownload(fh, request)
done = False
while done is False:
    status, done = downloader.next_chunk()
    print("Download %d%%." % int(status.progress() * 100))
os.chmod(filename, 600)

Download 100%.
Download 100%.


In [12]:
# hide
filename = "/content/.reed/creds.json"
with open(filename, 'r') as f:
    creds = json.load(f)
    
api_key=creds['API_KEY']

## Intro

Visualisation is an essential part of working with data. Whether we are delivering regular reports to decision-makers, presenting the results of a regression model to a room of stakeholders, or creating a real-time dashboard for the wider business, we are using data to tell a story, and visuals bring these stories to life.

In this post we are going to explore a number of ways to visualise text data in the context of analysing job posts for three data roles: data scientists, data engineers and data analysts. These roles have a lot in common but are fundamentally different; will the content of the job descriptions for these roles reflect the differences between them? And how can we visualise these differences?

## Getting the data

First we need to collect some job descriptions to analyse. There are plenty of job search APIs out there to choose from and we also have the option to scrape directly from job websites. I ended up using the Reed developer API because of the simple process of signing up for credentials. I wrote a python wrapper for the API which I will be using to extract a collection of relevant job posts; you can find it [here](https://pypi.org/project/reed/).

It's a simple process to fetch the jobs matching a keywords search query. The response is a list of json objects which in this case are identical to python dictionaries so we can easily place the data in a pandas dataframe. 

In [13]:
import pandas as pd
from reed import ReedClient

client = ReedClient(api_key=api_key)
search_params = {
    'keywords': 'data+scientist|data+engineer|data+analyst', 
    'resultsToTake': 600
}
response = client.search(**search_params)
df = pd.DataFrame(response)
df.shape

(524, 15)

Our search returned 524 jobs in total; let's clean up the job titles and see what we have.

In [14]:
def clean_title(title_str): 
    return title_str.lower().strip().replace('  ', ' ')

df['jobTitle'] = [clean_title(x) for x in df.jobTitle]
df.groupby('jobTitle').size().sort_values(ascending=False).head(10)

jobTitle
data analyst                   62
data engineer                  46
data scientist                 43
senior data scientist          19
senior data engineer           11
big data engineer               5
lead data scientist             5
lead data engineer              4
data scientist - manchester     4
graduate data analyst           4
dtype: int64

We're going to remove all the posts which don't have one of the three most common titles; what we're left with should be broadly similar in terms of representing mid-level roles within each category. 

In [15]:
accepted_titles = ['data scientist', 'data engineer', 'data analyst']
df = df[[x in accepted_titles for x in df.jobTitle]].set_index('jobId')

The job search API gives a truncated version of the job descriptions so if we want the complete text we'll have to take each individual job id and pull down the details using the job details function one-by-one. To make this process easier I created a dictionary with the three job titles as keys and the items consisting of the list of ids corresponding to each title.

In [16]:
groups = df.groupby('jobTitle').groups
groups.keys()

dict_keys(['data analyst', 'data engineer', 'data scientist'])

Now we can loop over the dictionary pulling down the description from each job in turn. There's another complication in that the job descriptions are returned as html documents. We're only interested in the text data so we're going to have the parse the html to extract the information we want. We can wrap this process inside a function which we call for each id in our dictionary. 

In [17]:
def get_job_desc(job_type, job_id, client, html_parser):
    desc_html = client.job_details(job_id)['jobDescription']
    parser.feed(desc_html)
    desc_str = parser.return_data()
    # reset parser string
    parser.string = ''
    return dict(job=job_type, job_id=job_id, desc=desc_str)
    
parser = MyHTMLParser()
job_descriptions = []
for title in groups:
    for id in groups[title]:
        job_desc_dict = get_job_desc(title, id, client, parser)
        job_descriptions.append(job_desc_dict)
        
df = pd.DataFrame(job_descriptions)
df.head()

job    job_id                                               desc
0  data analyst  40501878  Data Analyst | Healthcare / MedTech | SQL, Tab...
1  data analyst  40532763  Data Analyst  £35,000 - £40,000 + Benefits  Lo...
2  data analyst  40500203  A well established Telecommunications company ...
3  data analyst  40500186  A well established Telecommunications company ...
4  data analyst  40531962  My client, a multinational, speciality insuran...

In order to visualise the content of the job descriptions we need a numerical representation of the text. One way to do this is with a bag-of-words approach, which consists of separating the text in our documents into tokens and counting the appearances of each token in each document. Before we do this there are some errands we need to run first. Sklearn's Countvectorizer class performs the tokenization step for us, but we need to be careful when using it. By default it splits the text on punctuation marks and discards tokens less than two characters long. For this reason, the word 'Ph.D' which is frequently used would be split into the two meaningless tokens 'Ph' and 'D', for example. The important term 'R' denoting the programming language would also be discarded. We can remedy these issues and others similar using pandas string methods. At the same time we'll remove any rows with duplicate job descriptions.

In [18]:
# hide
filename = "/content/data/job_data.csv"
df = pd.read_csv(filename, index_col=0)

In [19]:
df = df[~df.duplicated('desc')]

cicd_pat = "([Cc][Ii]/[Cc][Dd])"
phd_pat = "[Pp][Hh].?[Dd]"
r_pat = '(\sR\W)'

df['desc'] = (df.desc.str.replace(',000', 'k')
             .str.replace('\xa0', ' ')
             .str.replace(phd_pat, 'PHD')
             .str.replace(r_pat, ' RStudio ')
             .str.replace(cicd_pat, ' CICD')
             .str.replace('Modis', ''))

Now we're finally at a stage where our data is ready to be analysed. After whittling the results down to mid-level roles we were left with 133 unique jobs with a roughly even split between the three roles. 

When we fit the CountVectorizer transform to an array containing each job description what we get is a document-term matrix: the rows are the job descriptions and the columns are all the words which were found in the collection of documents. The contents of the matrix are the word frequencies. 

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
count_vectorizer = CountVectorizer(stop_words='english')
text_data = np.array(df['desc'])
count_matrix = count_vectorizer.fit_transform(text_data)
count_matrix_df = pd.DataFrame(count_matrix.toarray(), columns=count_vectorizer.get_feature_names())
count_matrix_df

00  000  025  04  06  075  ...  year  years  yields  yoga  yorkshire  zachman
0     0    0    0   0   0    0  ...     0      0       0     0          0        0
1     0    0    0   0   0    0  ...     0      0       0     0          0        0
2     0    0    0   0   0    0  ...     0      0       0     0          0        0
3     0    0    0   0   0    0  ...     0      0       0     0          0        0
4     0    0    0   0   0    0  ...     0      0       0     0          0        0
..   ..  ...  ...  ..  ..  ...  ...   ...    ...     ...   ...        ...      ...
128   0    0    0   0   0    0  ...     0      1       0     0          0        0
129   0    0    0   0   0    0  ...     0      0       0     0          0        0
130   0    0    0   0   0    0  ...     0      1       0     0          0        0
131   0    0    0   0   0    0  ...     0      0       0     0          0        0
132   0    0    0   0   0    0  ...     0      1       0     0          0        0

[133 rows x 3741 columns]

How can we use this implementation to visualise the three different roles? A simple option is to look at the most frequent words across the three classes: if we group the documents into the three roles, sum the word frequencies and sort them in descending order we'll see which words appear the most in each.

In [21]:
# hide_input
from plotly.subplots import make_subplots
import plotly.graph_objects as go

count_matrix_df['job'] = df.job
counts_by_job = (count_matrix_df.groupby('job')
                 .sum()
                 .transpose()
                 .reset_index()
                 .rename(columns={'index': 'word'}))

da_top = counts_by_job[['word', 'data analyst']].sort_values('data analyst', ascending=False)[:20]
de_top = counts_by_job[['word', 'data engineer']].sort_values('data engineer', ascending=False)[:20]
ds_top = counts_by_job[['word', 'data scientist']].sort_values('data scientist', ascending=False)[:20]

colours_dict = {'data analyst': '#636efa', 'data scientist': '#00cc96', 'data engineer': '#EF553B'}

fig = make_subplots(rows=1, 
                    cols=3, 
                    subplot_titles=("DS top words", "DE top words", "DA top words"),
                    x_title='frequency'
                )

fig.add_trace(go.Bar(
    y=ds_top['word'],
    x=ds_top['data scientist'],
    orientation='h',
    name='ds',
    marker=dict(color=colours_dict['data scientist'])),
    row=1, col=1
)

fig.add_trace(go.Bar(
    y=de_top['word'],
    x=de_top['data engineer'],
    orientation='h',
    name='de',
    marker=dict(color=colours_dict['data engineer'])),
    row=1, col=2
)

fig.add_trace(go.Bar(
    y=da_top['word'],
    x=da_top['data analyst'],
    orientation='h',
    name='da',
    marker=dict(color=colours_dict['data analyst'])),
    row=1, col=3
)

fig.update_layout(
    yaxis1=dict(autorange="reversed"),
    yaxis2=dict(autorange="reversed"),
    yaxis3=dict(autorange="reversed"),
    height=600, width=1000,
    showlegend=False
)
fig.show()

We can spot a few themes such as the emphasis on machine learning in DS job descriptions while SQL and reporting feature prominently in DA posts. The plot corresponding to Data engineer roles is not as insightful. The appearance of Python, SQL and Azure are promising, but no sign of ETL pipelines and a lower prevalance of cloud-based systems than we'd expect. And why does scientist turn up so much? Across the three plots we can also see another problem with this simple word frequency approach: the most insightful terms are diluted by a saturation of words which are common across the three roles and thus essentially useless, such as 'data', 'experience', 'role' and 'team'. 

We need a metric which highly ranks frequent words but suppresses words which appear in lots of the documents in our corpus. Luckily for us a metric exists which does exactly as we require. TFIDF or term-frequency inverse-document frequency takes the term frequency of a word in a document and multiplies it by the inverse document freqency of that word, essentially damping those words which appear across a large proportion of the documents. Again sklearn has our back and can perform this process for us.

In [22]:
from sklearn.feature_extraction.text import TfidfVectorizer
max_df = 0.6
min_df = 2
vectorizer  = TfidfVectorizer(stop_words='english', max_df=max_df, min_df=min_df)
text = np.array(df['desc'])
tfidf_matrix = vectorizer.fit_transform(text)
tfidf = pd.DataFrame(tfidf_matrix.toarray(), columns=vectorizer.get_feature_names())
tfidf

00   08   10  100  1070906  ...   writing  written   x4  year     years
0    0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.000000
1    0.0  0.0  0.0  0.0      0.0  ...  0.086106      0.0  0.0   0.0  0.000000
2    0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.000000
3    0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.000000
4    0.0  0.0  0.0  0.0      0.0  ...  0.100455      0.0  0.0   0.0  0.000000
..   ...  ...  ...  ...      ...  ...       ...      ...  ...   ...       ...
128  0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.044717
129  0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.000000
130  0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.093065
131  0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.000000
132  0.0  0.0  0.0  0.0      0.0  ...  0.000000      0.0  0.0   0.0  0.044297

[133 rows x 2038 columns]

As before the TfidfVectorizer returns a term-document matrix, but instead of consisting of word frequencies we have tfidf values for each term-document pair. The parameters 'min_df' and 'max_df' are constraints on the document frequency of the words in our vocabulary. We have used a 'min_df' of 2, so any words that do not appear in at least two documents are thrown out. 'Max_df' is a float which corresponds to a proportion of the documents, so here if any words appear in more than 60 percent of the documents they too are discarded. In the same way as before, we can sum the tfidf values within the three classes and plot the highest aggregate values to hopefully extract the most important terms to characterise the three job types.

In [23]:
# hide_input
tfidf_with_job = tfidf.copy()
tfidf_with_job['job_type'] = df.job
counts_by_job = (tfidf_with_job.groupby('job_type')
                 .sum()
                 .transpose()
                 .reset_index()
                 .rename(columns={'index': 'word'}))

da_top = counts_by_job[['word', 'data analyst']].sort_values('data analyst', ascending=False)[:20]
de_top = counts_by_job[['word', 'data engineer']].sort_values('data engineer', ascending=False)[:20]
ds_top = counts_by_job[['word', 'data scientist']].sort_values('data scientist', ascending=False)[:20]

colours_dict = {'data analyst': '#636efa', 'data scientist': '#00cc96', 'data engineer': '#EF553B'}

fig = make_subplots(rows=1, 
                    cols=3, 
                    subplot_titles=("DS top words", "DE top words", "DA top words"),
                    x_title='tfidf score'
                )

fig.add_trace(go.Bar(
    y=ds_top['word'],
    x=ds_top['data scientist'],
    orientation='h',
    marker=dict(color=colours_dict['data scientist'])),
    row=1, col=1
)

fig.add_trace(go.Bar(
    y=de_top['word'],
    x=de_top['data engineer'],
    orientation='h',
    marker=dict(color=colours_dict['data engineer'])),
    row=1, col=2
)

fig.add_trace(go.Bar(
    y=da_top['word'],
    x=da_top['data analyst'],
    orientation='h',
    marker=dict(color=colours_dict['data analyst'])),
    row=1, col=3
)

fig.update_layout(
    yaxis1=dict(autorange="reversed"),
    yaxis2=dict(autorange="reversed"),
    yaxis3=dict(autorange="reversed"),
    height=600, width=1000,
    showlegend=False
)
fig.show()

This is a big improvement over our first attempt. There are far fewer generic terms because of the introduction of the inverse-document weighting and the document frequency constraints. The data engineering terms are also closer to our expectations than they were before. 

What are the key insights we can take from these plots? For data scientists machine learning is the most distinguishing feature. R and Python both appear with the latter ranked slightly higher, at least in this sample. We can also see that doctorates are important in data scientist posts but not in the other two roles.

Our picture of data engineering posts is clearer from this version of the plot. Cloud technologies, ETL pipelines and databases feature heavily and python seems to be the scripting language of choice within this sample of posts. The impression given by this visual is that tools and technologies are to data engineering what algorithms and models are to data science.

Finally, data analyst positions appear to be characterised more by aptitudes and skills than by specific technologies or technical backgrounds. SQL and excel feature heavily but apart from these tools the key terms seem to describe an general analytical mindset and an ability to support other business functions through reporting and analysis.

## Plotting job descriptions via dimensionality reduction

So far we've used the TF-IDF vectors to gain insight into the key words that describe the different data roles. We haven't, however, used any unsupervised learning as we promised at the beginning. Let's sort that out now. 

Using TF-IDF we have created a 2038-dimensional representation of the job descriptions in our sample. As it is a bit of a struggle to visualise anything in more than three dimensions, it is common practice to manipulate high-dimensional data into a two or three dimensional space so that it can be visualised and structural patterns in the data can be more easily found. In this case we are going to use Principal Component Analysis, or PCA for short, to project our data onto the two dimensions which retain the highest variability. This will allow us to create a scatterplot of the data.

In [24]:
from sklearn.decomposition import PCA
pca = PCA(n_components=2)
X = pca.fit_transform(tfidf)
X_df = pd.DataFrame(X)
X_df['job'] = df.job.values
X_df = X_df.rename(columns={0: 'pc_0', 1: 'pc_1'})
X_df.sample(5)

pc_0      pc_1             job
102  0.210452 -0.119038  data scientist
6   -0.135975 -0.110068    data analyst
55   0.111871  0.328581   data engineer
81  -0.068974  0.126661   data engineer
31  -0.205418 -0.054962    data analyst

In [25]:
pca.explained_variance_

array([0.03001154, 0.02439249])

Using only a two dimensional representation of the TF-IDF vectors we retain just over half of the variance in our data. The following is a scatterplot of the first two principle components; each point is a job description.

In [26]:
import plotly.express as px
fig = px.scatter(X_df, x="pc_0", y="pc_1", color="job", hover_data=["pc_0", "pc_1", X_df.index])
fig.update_layout(title='first two principle components of tfidf matrix')
fig.show()

Remarkably, the three different job titles are almost perfectly separated into clusters in this two dimensional representation. This is significant because the PCA algorithm has no idea that our dataset contains three distinct classes, but by simply projecting the data onto the coordinate axes retaining the most variability the algorithm has managed to almost perfectly separate them. We can go one step further and demonstrate that the three classes are well-separated by applying k-means clustering first two principle components. The result is the following.

In [27]:
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=3)
# X is the data matrix of the first two PC's
kmeans.fit(X)
y_kmeans = kmeans.predict(X)
centers = kmeans.cluster_centers_

In [28]:
# hide_input
fig = make_subplots(rows=2, cols=1, shared_xaxes=False, vertical_spacing=0.15,
                   subplot_titles=("actual job titles","k means clustered job titles"))

fig1 = px.scatter(X_df, x="pc_0", y="pc_1", color="job", hover_data=["pc_0", "pc_1", X_df.index])
trace1, trace2, trace3 = fig1['data'][0], fig1['data'][1], fig1['data'][2]
trace1['name'] = 'data analyst'
trace2['name'] = 'data engineer'
trace3['name'] = 'data scientist'

plotly_colours = px.colors.qualitative.Plotly
c = [plotly_colours[x] for x in y_kmeans]

fig2 = px.scatter(X_df, x='pc_0', y='pc_1', color=c, hover_data=['job', 'pc_0', 'pc_1', X_df.index])
trace4, trace5, trace6 = fig2['data'][0], fig2['data'][1], fig2['data'][2]
trace4['name'] = 'data analyst'
trace5['name'] = 'data engineer'
trace6['name'] = 'data scientist'

temp_str = '<br>job=%{customdata[0]}' + \
            '<br>pc_0=%{customdata[1]}' + \
            '<br>pc_1=%{customdata[2]}' + \
            '<br>index=%{customdata[3]}' + \
            '<extra></extra>'

trace4['hovertemplate'] = temp_str
trace5['hovertemplate'] = temp_str
trace6['hovertemplate'] = temp_str

trace4['legendgroup'] = 'job=data analyst'
trace4['showlegend'] = False
trace5['legendgroup'] = 'job=data engineer'
trace5['showlegend'] = False
trace6['legendgroup'] = 'job=data scientist'
trace6['showlegend'] = False

fig.add_trace(trace1, row=1, col=1)
fig.add_trace(trace2, row=1, col=1)
fig.add_trace(trace3, row=1, col=1)
fig.add_trace(trace4, row=2, col=1)
fig.add_trace(trace5, row=2, col=1)
fig.add_trace(trace6, row=2, col=1)

fig.add_trace(go.Scatter(
    x=centers[:, 0],
    y=centers[:, 1],
    name='cluster centers',
    mode='markers',
    marker=dict(size=[15, 15, 15],
                color=[2, 2, 2],
                opacity=0.7)),
    row=2, col=1
)

# fig['data'][6]['showlegend'] = False

fig.update_layout(height=800, legend=dict(x=1, y=0.5))
fig.layout['xaxis']['title'] = 'pc_0'
fig.layout['xaxis2']['title'] = 'pc_0'
fig.layout['yaxis']['title'] = 'pc_1'
fig.layout['yaxis2']['title'] = 'pc_1'
fig.show()

Let's recap what we've achieved here. We started with a collection of raw text documents, each one a description of a job from one of three different titles. Using term-frequency inverse-document-freqency we converted each description into a fixed length vector of TF-IDF values. Then using PCA we projected our new 2038-dimensional data matrix onto the two coordinate axes of highest variability so we could produce a scatterplot of the data. The result was a clear visual separation into three clusters corresponding to the three job titles, which we have now shown are nearly perfectly recreated using a simple clustering algorithm.

The cool part about this is not that we have accurately classified each TF-IDF vector with its job title. There are lots of supervised approaches that would likely be more effective if that were our goal. No, the interesting part is thinking about how these simple unsupervised methods can be applied to problems where we *don't* have access to labelled data as we did here. For example, given an unlabelled text dataset we could take the same steps, compute the TF-IDF values, plot the first two principle components and look for clusters. Then by comparing the most import words in each cluster via the TF-IDF values we might uncover patterns in the data that we weren't aware of. What clusters might we find if we applied this to restaurant reviews? Or house listings?

Hopefully some of these ideas for visualising and investigating text data have piqued your interest and you've learned some of the differences between data science, engineering and analysis roles along the way.  